flux/flux-bootstrap.sh

```diff
@@ -0,0 +1,120 @@
+#!/bin/bash
+
+# Override with the `flux.env.KUBECONFIG` key if necessary
+export KUBECONFIG=/etc/rancher/k3s/k3s.yaml
+
+# Seconds
+short=5
+long=900
+
+configmap=flux-bootstrap
+
+info () {
+  echo $'\e[36mINFO\e[0m: ' "$1"
+}
+
+warn() {
+  echo $'\e[33mWARN\e[0m:' "$1"
+}
+
+error() {
+  echo $'\e[31mERROR\e[0m:' "$1"
+}
+
+cleanup() {
+  info "Removing bootstrap configmap"
+  timeout $short kubectl delete configmap -n default $configmap
+}
+
+# Don't bootstrap if the cluster is already bootstrapped
+if flux version &>/dev/null; then
+  info "Flux is already bootstrapped, exiting..."
+  exit 0
+fi
+
+# Determine what VCS we need to bootstrap
+for vcs in bitbucket_server git github gitlab; do
+  if [[ $(kairos-agent config get flux.$vcs 2>/dev/null) != "null" ]]; then
+    version_control=$vcs
+    break
+  fi
+done
+
+if [[ "${version_control}x" == "x" ]]; then
+  error "Unable to determine what version control provider to use, exiting..."
+  exit 1
+fi
+
+# Get flux envs and settings for our VCS
+mapfile -t envs < <(kairos-agent config get "flux.env" 2>/dev/null)
+mapfile -t args < <(kairos-agent config get "flux.$version_control" 2>/dev/null)
+declare -a cmdline
+
+for setting in "${envs[@]}"; do
+  if [[ $setting != "null" ]]; then
+    env=$(echo "$setting" | cut -d: -f1)
+    value=$(echo "$setting" | cut -d':' -f2 | sed -E 's/^ //g')
+    if [[ "${value}x" != "x" ]]; then
+      export "$env"="$value"
+    fi
+  fi
+done
+
+# Set commandline args
+for setting in "${args[@]}"; do
+  if [[ $setting != "null" ]]; then
+    arg=$(echo "$setting" | cut -d: -f1)
+    value=$(echo "$setting" | cut -d':' -f2 | sed -E 's/^ //g')
+    if [[ "${value}x" != "x" ]]; then
+      cmdline+=("--$arg" "$value")
+    fi 
+  fi
+done
+
+# Try to bootstrap Flux for 30 minutes, sleep 15 seconds between attempts
+minutes=30
+sleep=15
+retry_attempt=1
+total_attempts=$(( minutes * 60 / sleep ))
+active="false"
+
+if [[ "${#cmdline[@]}" -eq 0 ]]; then
+  info "Flux was not configured in cloud-config, not bootstrapping"
+  exit 0
+else
+  while [[ $retry_attempt -le $total_attempts ]]; do
+    if [[ "$active" != "true" ]]; then
+      # Ensure only one host tries to bootstrap, whichever makes the configmap first
+      if ! timeout $short kubectl version &> /dev/null; then
```

Comment:
```localhost:/home/kairos # kubectl version && echo "ok"
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.25.11+k3s1
WARNING: version difference between client (1.28) and server (1.25) exceeds the supported minor version skew of +/-1
ok
```

but 

```
localhost:/home/kairos # journalctl -f -u flux-bootstrap
Oct 19 10:56:51 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:56:51 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 19 (of 120) failed, retrying in 15 seconds
Oct 19 10:57:06 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:57:06 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 20 (of 120) failed, retrying in 15 seconds
Oct 19 10:57:21 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:57:21 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 21 (of 120) failed, retrying in 15 seconds
Oct 19 10:57:36 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:57:36 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 22 (of 120) failed, retrying in 15 seconds
Oct 19 10:57:51 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:57:51 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 23 (of 120) failed, retrying in 15 seconds
```

I'm not sure what's going on. I will debug more.

Reply:
I added these lines right before the check:

```
      echo $KUBECONFIG
      whoami
      kubectl version

```

and this is the output:

```
localhost:/home/kairos # /usr/local/bin/flux-bootstrap.sh
/home/csagan/.kube/config
root
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
The connection to the server localhost:8080 was refused - did you specify the right host or port?

```

Reply:
something changes the value of KUBECONFIG :thinking:

Reply:
I'm stupid :facepalm: . I copied and pasted the config you provided and that sets it. Ignore the noise.